From 2d60b40cfa46df70cf5a918a3c0208faa3643168 Mon Sep 17 00:00:00 2001
From: Austin Mayes <austin@avicus.net>
Date: Mon, 31 Dec 2018 18:15:40 -0600
Subject: [PATCH] Store remote IP in error log records


diff --git a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
index a66099fc..76a5c798 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
@@ -9,6 +9,7 @@ import io.netty.handler.timeout.ReadTimeoutException;
 import java.io.IOException;
 import java.net.InetSocketAddress;
 import java.util.logging.Level;
+import java.util.logging.LogRecord;
 import net.md_5.bungee.api.ProxyServer;
 import net.md_5.bungee.connection.CancelSendSignal;
 import net.md_5.bungee.connection.InitialHandler;
@@ -156,7 +157,10 @@ public class HandlerBoss extends ChannelInboundHandlerAdapter
                     } );
                 } else
                 {
-                    ProxyServer.getInstance().getLogger().log( Level.SEVERE, handler + " - encountered exception", cause );
+                    final LogRecord record = new LogRecord(Level.SEVERE, handler + " - encountered exception" );
+                    record.setThrown( cause );
+                    record.setParameters( new Object[]{ ctx.channel().remoteAddress() } );
+                    ProxyServer.getInstance().getLogger().log( record );
                 }
             }
 
-- 
2.20.1

